File: Programas/NoInformados/Rompecabezas/2. Busqueda en Amplitud/C++ BFS/Rompecabezas.cpp
#include "./Header/Rompecabezas.h"
#include "./Header/Gestor.h"

Rompecabezas::Rompecabezas(Rompecabezas *nodopadre, char accion, int profundidad){
    this->nodopadre = nodopadre;
    this->profundidad = profundidad;
    if (nodopadre == nullptr){
        this->tablero = 0;
        Gestor g;
        for (int valor : g.elementos){
            this->tablero = this->tablero | ((unsigned long long) valor << (unsigned long long)(4 * (valor-1)));
        }
        this->cero = 8;
    }
    else{
        this->tablero = nodopadre->tablero;
        this->cero = nodopadre->cero;
        if (accion != ' '){
            mover(accion);
        }
    }
}

void Rompecabezas::mostrarRompecabezas(){
    for (int i = 0; i < 9; i++){
        std::cout << Gestor::obtenerCelda( i , this->tablero ) << " ";
        if( i % 3 == 2 )
            std::cout << "\n";
    }
    std::cout << "\n";
}

void Rompecabezas::mover(char accion){
    if(accion == 'N') 
        Gestor::mover_celda_arriba( this );
    if(accion == 'E') 
        Gestor::mover_celda_derecha( this );
    if(accion == 'S') 
        Gestor::mover_celda_abajo( this );
    if(accion == 'O') 
        Gestor::mover_celda_izquierda( this );
}

bool Rompecabezas::compararCeldas(Rompecabezas *rompecabezas){
    return this->tablero == rompecabezas->tablero;
}

void Rompecabezas::modificarCeldas(int celdas[]){
    this->tablero = (unsigned long long) 0;
    for( int i = 8 ; i >= 0 ; i-- ){
        this->tablero = this->tablero << 4 | celdas[i];
        if( celdas[i] == 0 )
            this->cero = i;
    }
}

std::vector<Rompecabezas*> Rompecabezas::expandirNodos(){
    std::vector<Rompecabezas*> nodos;
    Gestor g;
    for( char movimiento : g.movimientos_validos[this->cero] ){
        Rompecabezas * hijo = new Rompecabezas( this , movimiento , this->profundidad+1 );
        if( this->nodopadre != nullptr ){
            if( hijo->tablero != this->nodopadre->tablero )
                nodos.push_back( hijo );
        }else{
            nodos.push_back( hijo );
        }
    }
    return nodos;
}

File: Programas/NoInformados/Rompecabezas/1. Busqueda en Profundidad v1.0.0/JAVA DFS/Rompecabezas.java
import java.util.ArrayList;

public class Rompecabezas {

  public int cero;
  public long tablero;
  public Rompecabezas nodoPadre;
  public int profundidad;

  public Rompecabezas(Rompecabezas nodoPadre, String accion, int profundidad) {
    this.nodoPadre = nodoPadre;
    this.profundidad = profundidad;
    if (nodoPadre == null) {
      this.tablero = 0;
      for (int valor : Gestor.elementos) {
        this.tablero = this.tablero | ((long) valor << (long) (4 * (valor - 1)));
      }
      this.cero = 8;
    } else {
      this.tablero = nodoPadre.tablero;
      this.cero = nodoPadre.cero;
      if (accion != null)
        mover(accion);
    }
  }

  public void mover(String accion) {
    if (accion == "N")
      Gestor.mover_celda_arriba(this);
    else if (accion == "E")
      Gestor.mover_celda_derecha(this);
    else if (accion == "S")
      Gestor.mover_celda_abajo(this);
    else if (accion == "O")
      Gestor.mover_celda_izquierda(this);
  }

  public void mostrarRompecabezas() {
    for (int i = 0; i < 9; i++) {
      System.out.print(Gestor.obtenerCelda(i, this.tablero) + " ");
      if (i % 3 == 2)
        System.out.println();
    }
    System.out.println();
  }

  public Boolean compararCeldas(Rompecabezas otro) {
    return otro.tablero == this.tablero;
  }

  public void revolverCeldas(int numeroMovimientos) {
    for (int i = 0; i < numeroMovimientos; i++) {
      String[] movimientos_validos_cero = Gestor.movimientos_validos.get(
          this.cero);
      int posicion_movimientos = (int) (Math.random() * (movimientos_validos_cero.length - 1 - 0 + 1) + 0);
      mover(movimientos_validos_cero[posicion_movimientos]);
    }
  }

  public ArrayList<Rompecabezas> expandirNodos() {
    ArrayList<Rompecabezas> nodos = new ArrayList<>();
    for (String movimiento : Gestor.movimientos_validos.get(this.cero)) {
      Rompecabezas nodo = new Rompecabezas(
          this,
          movimiento,
          this.profundidad + 1);
      if (this.nodoPadre != null) {
        if (nodo.tablero != this.nodoPadre.tablero)
          nodos.add(nodo);
      } else {
        nodos.add(nodo);
      }
    }
    return nodos;
  }
}
